code/modules/mob/living/carbon/xenomorph/castes/praetorian/abilities_praetorian.dm

```diff
@@ -197,3 +197,187 @@ GLOBAL_LIST_INIT(acid_spray_hit, typecacheof(list(/obj/structure/barricade, /obj
+
+	owner.add_movespeed_modifier(MOVESPEED_ID_PRAETORIAN_DANCER_DODGE_SPEED, TRUE, 0, NONE, TRUE, speed_buff)
+	owner.allow_pass_flags |= (PASS_MOB|PASS_XENO)
+	owner.pass_flags |= (PASS_MOB|PASS_XENO)
+	RegisterSignal(owner, COMSIG_MOVABLE_MOVED, PROC_REF(on_move))
+	addtimer(CALLBACK(src, PROC_REF(remove_effects)), duration)
+
+	succeed_activate()
+	add_cooldown()
+
+/// Automatically bumps living non-xenos if bump attacks are on.
+/datum/action/ability/xeno_action/dodge/proc/on_move(datum/source)
+	if(!isxeno(owner))
+		return FALSE
+	var/mob/living/carbon/xenomorph/xeno_owner = owner
+	if(xeno_owner.stat == DEAD)
+		return FALSE
+	var/datum/action/bump_attack_toggle/bump_attack_action = xeno_owner.actions_by_path[/datum/action/bump_attack_toggle]
+	if(bump_attack_action == null || bump_attack_action.attacking) // Bump attacks are off if attacking is true, apparently.
+		return FALSE
+	if(TIMER_COOLDOWN_CHECK(src, COOLDOWN_BUMP_ATTACK))
+		return FALSE
+
+	var/turf/current_turf = get_turf(owner)
+	for(var/mob/living/living_mob in current_turf)
+		if(living_mob.stat == DEAD)
+			continue
+		if(isxeno(living_mob))
+			var/mob/living/carbon/xenomorph/xenomorph_mob = living_mob
+			if(owner.issamexenohive(xenomorph_mob))
+				continue
+		xeno_owner.Bump(living_mob)
+		return
+
+/// Removes the movespeed modifier and various pass_flags that was given by the dodge ability.
+/datum/action/ability/xeno_action/dodge/proc/remove_effects()
+	owner.balloon_alert(owner, "Dodge inactive!")
+
+	owner.remove_movespeed_modifier(MOVESPEED_ID_PRAETORIAN_DANCER_DODGE_SPEED)
+	owner.allow_pass_flags &= ~(PASS_MOB|PASS_XENO)
+	owner.pass_flags &= ~(PASS_MOB|PASS_XENO)
+	UnregisterSignal(owner, COMSIG_MOVABLE_MOVED)
+
+// ***************************************
+// *********** Impale
+// ***************************************
+/datum/action/ability/activable/xeno/impale
+	name = "Impale"
+	action_icon_state = "impale"
+	action_icon = 'icons/Xeno/actions/praetorian.dmi'
+	desc = "Impale a marine next to you with your tail for moderate damage. Marked enemies are impaled twice."
+	ability_cost = 100
+	cooldown_duration = 8 SECONDS
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_IMPALE,
+	)
+
+/datum/action/ability/activable/xeno/impale/can_use_ability(atom/A, silent = FALSE, override_flags)
+	. = ..()
+	if(!.)
+		return FALSE
+	if(!iscarbon(A))
+		if(!silent)
+			A.balloon_alert(owner, "cannot impale")
+		return FALSE
+	var/mob/living/carbon/carbon_target = A
+	if(!owner.Adjacent(carbon_target))
+		carbon_target.balloon_alert(owner, "too far")
+		return FALSE
+	if(carbon_target.stat == DEAD)
+		carbon_target.balloon_alert(owner, "already dead")
+		return FALSE
+
+/datum/action/ability/activable/xeno/impale/use_ability(atom/target_atom)
+	. = ..()
+
+	var/mob/living/carbon/xenomorph/xeno_owner = owner
+	var/mob/living/carbon/living_target = target_atom
+	var/buffed = living_target.has_status_effect(STATUS_EFFECT_DANCER_TAGGED)
```

Comment:
This will runtime if you click something that isnt a person

Reply:
can_use_ability checks if target is carbon. not possible to runtime or?